Made updates to the final task. Clarified scores, included history se… (#1677)
rolling-scopes-school/tasks: react/modules/tasks/final.md

### General requirements

- Errors (Not found route, unhandled rejection, etc) should be displayed in a user-friendly format (toast, pop-up, or something like that - up to your decision).

> - Internationalization (i18n) - at least 2 languages. You should be able to change the language by clicking on the toggler/select in the header.

Reviewer: ```suggestion
- Internationalization (i18n) - at least 2 languages. User should be able to change the language by clicking on the toggler/select in the header.
```

#### Error handling

Application should gracefully handle errors and show meaningfull messages to the user.
Please, distiungush between errors on the application level and the valid HTTP error code.

> CORS issues or network timeouts should be shown as errors, whereas 4xx and 5xx HTTP response codess should be displayed on the response section.

Reviewer: ```suggestion
CORS issues or network timeouts should be shown as errors, whereas 4xx and 5xx HTTP response codes should be displayed in the response section.
```

curl -d '{"title":"fakeTitle","userId":1,"body":"fakeMessage"}' https://jsonplaceholder.typicode.com/posts
```

Headers, specified in the editor, should be provided as url query parameters (mind that you need to url encode headers values):
`http://localhost:5137/POST/aHR0cHM6Ly9qc29ucGxhY2Vob2xkZXIudHlwaWNvZGUuY29tL3Bvc3Rz/eyJ0aXRsZSI6ImZha2VUaXRsZSIsInVzZXJJZCI6MSwiYm9keSI6ImZha2VNZXNzYWdlIn0=?Content-Type=application%2Fjson`

### GraphiQL

- This route should be private.
- Header should be visible.
- Endpoint (url) input.

> - SDL endpoint (url) input which will be used for the documentation. By default will duplicate the value provided in the main url input, and will add "?sdl" to the end. Use should be able to modify this url if the api uses another endpoint for the documentation.

Reviewer: ```suggestion
- SDL endpoint (url) input which will be used for the documentation. By default will duplicate the value provided in the main url input, and will add "?sdl" to the end. User should be able to modify this url if the api uses another endpoint for the documentation.
```

- Headers editor section (If you are making a CORS request each added header should be supported on the backend, please, consider that). You CAN reuse the one from the RESTfull client.
- Documentation section, should be visible only when the app receives a successfull response with the schema definition from the API.

#### Routing on GraphiQL client

Similar to the RESTfull client, the request infromation should be provided via the url, with the only exception - all the GraphQL queries are POST queries, so to distinguish it from the usual REST queries, you should use GRAPHQL as a first route parameter:
`http://localhost:5137/GRAPHQL/{endpointUrlBase64encoded}/{bodyBase64encoded}?header1=header1value&header2=header2value...`

### History

- This route should be private.
- If there are no requests in the local storage, show message to the user, e.g. "You haven't executed any requests yet", "It's empty here. Try those options:" and give links to the RESTfull and to the GraphiQL clients.
- All the requests should be saved in the local storage on submit, history section should show requests sorted by the time of their execution.

> - History section should display request using links, on clicking on the link, user should be havigated to the respective section (RESTfull client or GraohiQL).

Reviewer: ```suggestion
- History section should display request using links, on clicking on the link, user should be navigated to the respective section (RESTfull client or GraphiQL).
```

- [ ] Functional read-only response section, with information about HTTP status and the code. - **30 mpoints**
- [ ] Method selector, shows all the valid HTTP verbs, value is provided in the url on change. - **10 points**
- [ ] Input for the url, entered value is provided in base64-encoded way on change. - **15 points**
- [ ] Variables section that can shown or hidden, specified variables are included in the body. - **15 points**
- [ ] Headers section, value is provided in the url on header add/change. - **20 points**

### GraphiQL route - max 80 points

- [ ] Functional editor enabling query editing and prettifying, request body provided in the url as base64-encoded on focus out. - **35 points**
- [ ] Read-only response section, with information about HTTP status and the code, reused from the RESTfull client. - **5 points**
- [ ] Operational documentation explorer, visible _only_ upon successful SDL request. - **20 points**
- [ ] Variables section that can shown or hidden, specified variables are included in the body. - **10 points**
- [ ] Header section that can be shown or hidden, value is provided in the url on header add/change. - **10 points**

> ### History route = max 50 points

Reviewer: ```suggestion
### History route - max 50 points
```

> - [ ] History shows informational message with links to the clients when there is no requests in the local storage. - **10 points**

Reviewer: ```suggestion
- [ ] History shows informational message with links to the clients when there are no requests in the local storage. - **10 points**
```
